File: iOS 8 by Tutorials/Objective-C/Chapter 12 - Extensions - Photo Editing/Imgvue/Imgvue Filters/PhotoEditingViewController.m
/*
 * Copyright (c) 2014 Razeware LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#import "PhotoEditingViewController.h"
#import <Photos/Photos.h>
#import <PhotosUI/PhotosUI.h>
#import "RWTImageFilterService.h"

@interface PhotoEditingViewController () <PHContentEditingController>

@property (strong) PHContentEditingInput *input;

@property (weak, nonatomic) IBOutlet UIImageView *imageView;
@property (weak, nonatomic) IBOutlet UIButton *addFilterButton;
@property (weak, nonatomic) IBOutlet UIButton *undoButton;

@property (strong, nonatomic) RWTImageFilterService *imageFilterService;
@property (strong, nonatomic) NSString *currentFilterName;
@property (strong, nonatomic) UIImage *filteredImage;

@end

@implementation PhotoEditingViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  
  self.imageFilterService = [[RWTImageFilterService alloc] init];
}

- (IBAction)undo:(id)sender {
  self.imageView.image = self.input.displaySizeImage;
  self.currentFilterName = nil;
  self.filteredImage = nil;
}

- (IBAction)addFilter:(UIButton *)sender {
  UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Add Filter" message:@"Choose a filter to apply to the image" preferredStyle:UIAlertControllerStyleActionSheet];
  
  [self.imageFilterService.availableFilters enumerateKeysAndObjectsUsingBlock:^(NSString *name, NSString *filter, BOOL *stop) {
    UIAlertAction *action = [UIAlertAction actionWithTitle:name style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
      self.filteredImage = [self.imageFilterService applyFilter:filter toImage:self.input.displaySizeImage];
      self.imageView.image = self.filteredImage;
      self.currentFilterName = filter;
    }];
    [alert addAction:action];
  }];
  
  UIAlertAction *noneAction = [UIAlertAction actionWithTitle:@"None" style:UIAlertActionStyleCancel handler:^(UIAlertAction *action) {
    [self undo:nil];
  }];
  
  [alert addAction:noneAction];
  [self presentViewController:alert animated:YES completion:nil];
}

- (BOOL)canHandleAdjustmentData:(PHAdjustmentData *)adjustmentData {
  // Inspect the adjustmentData to determine whether your extension can work with past edits.
  // (Typically, you use its formatIdentifier and formatVersion properties to do this.)
  return NO;
}
- (void)startContentEditingWithInput:(PHContentEditingInput *)contentEditingInput placeholderImage:(UIImage *)placeholderImage {
  // Present content for editing, and keep the contentEditingInput for use when closing the edit session.
  // If you returned YES from canHandleAdjustmentData:, contentEditingInput has the original image and adjustment data.
  // If you returned NO, the contentEditingInput has past edits "baked in".
  self.input = contentEditingInput;
  
  self.imageView.image = self.input.displaySizeImage;
  
}
- (void)finishContentEditingWithCompletionHandler:(void (^)(PHContentEditingOutput *))completionHandler {
  // Update UI to reflect that editing has finished and output is being rendered.
  
  if (self.currentFilterName == nil || self.input == nil) {
    [self cancelContentEditing];
    return;
  }
  
  // Render and provide output on a background queue.
  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
    // Create editing output from the editing input.
    PHContentEditingOutput *output = [[PHContentEditingOutput alloc] initWithContentEditingInput:self.input];
    NSData *archiveData = [NSKeyedArchiver archivedDataWithRootObject:self.currentFilterName];
    PHAdjustmentData *adjustmentData = [[PHAdjustmentData alloc] initWithFormatIdentifier:@"com.raywenderlich.imgvue-imgvue-filters" formatVersion:@"1.0" data:archiveData];
    output.adjustmentData = adjustmentData;
    
    UIImage *fullImage = [UIImage imageWithContentsOfFile:self.input.fullSizeImageURL.path];
    fullImage = [self.imageFilterService applyFilter:self.currentFilterName toImage:fullImage];
    
    NSData *jpegData = UIImageJPEGRepresentation(fullImage, 1.0);
    NSError *saveError;
    BOOL saved = [jpegData writeToURL:output.renderedContentURL options:NSDataWritingAtomic error:&saveError];
    if (saved) {
      completionHandler(output);
    } else {
      NSLog(@"An error occurred during save: %@", saveError);
      completionHandler(nil);
    }
  });
}
- (BOOL)shouldShowCancelConfirmation {
  // Returns whether a confirmation to discard changes should be shown to the user on cancel.
  // (Typically, you should return YES if there are any unsaved changes.)
  return NO;
}
- (void)cancelContentEditing {
  // Clean up temporary files, etc.
  // May be called after finishContentEditingWithCompletionHandler: while you prepare output.
}

@end

File: iOS 8 by Tutorials/Swift/Chapter 12 - Extensions - Photo Editing/Imgvue-Final/Imgvue Filters/PhotoEditingViewController.swift
/*
* Copyright (c) 2014 Razeware LLC
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*/

import UIKit
import Photos
import PhotosUI
import ImgvueKit

class PhotoEditingViewController: UIViewController, PHContentEditingController {
  
  var input: PHContentEditingInput?
  
  @IBOutlet weak var imageView: UIImageView!
  @IBOutlet weak var addFilterButton: UIButton!
  @IBOutlet weak var undoButton: UIButton!
  
  let filterService: ImageFilterService = ImageFilterService()
  var currentFilterName: String?
  var filteredImage: UIImage?
  
  
  @IBAction func undo(sender: UIButton) {
    if let input = input {
      imageView.image = input.displaySizeImage
      currentFilterName = nil
      filteredImage = nil
    }
  }
  
  @IBAction func addFilter(sender: UIButton) {
    let alertController =
    UIAlertController(title: "Add Filter",
      message: "Choose a filter to apply to the image",
      preferredStyle: .ActionSheet)
    for (name, filter) in filterService.availableFilters() {
      let action = UIAlertAction(title: name, style: .Default) {
        _ in
        if let input = self.input {
          self.filteredImage =
            self.filterService.applyFilter(filter, toImage:
              input.displaySizeImage)
          self.imageView.image = self.filteredImage
          self.currentFilterName = filter
        }
      }
      alertController.addAction(action)
    }
    let noneAction =
    UIAlertAction(title: "None", style: .Cancel) { action in
      if let input = self.input {
        self.imageView.image = input.displaySizeImage
        self.currentFilterName = nil
        self.filteredImage = nil
      }
    }
    alertController.addAction(noneAction)
    presentViewController(alertController, animated: true,
      completion: nil)
  }
  
  override func viewDidLoad() {
    super.viewDidLoad()
    // Do any additional setup after loading the view.
  }
  
  override func didReceiveMemoryWarning() {
    super.didReceiveMemoryWarning()
    // Dispose of any resources that can be recreated.
  }
  
  // MARK: - PHContentEditingController
  
  func canHandleAdjustmentData(adjustmentData: PHAdjustmentData?) -> Bool {
    // Inspect the adjustmentData to determine whether your extension can work with past edits.
    // (Typically, you use its formatIdentifier and formatVersion properties to do this.)
    return false
  }
  
  func startContentEditingWithInput(contentEditingInput: PHContentEditingInput?, placeholderImage: UIImage) {
    // Present content for editing, and keep the contentEditingInput for use when closing the edit session.
    // If you returned YES from canHandleAdjustmentData:, contentEditingInput has the original image and adjustment data.
    // If you returned NO, the contentEditingInput has past edits "baked in".
    
    input = contentEditingInput
    
    if let input = contentEditingInput {
      imageView.image = input.displaySizeImage
    }
  }
  
  func finishContentEditingWithCompletionHandler(completionHandler: ((PHContentEditingOutput!) -> Void)!) {
    // 1
    if currentFilterName == nil || input == nil {
      self.cancelContentEditing()
      return;
    }
    // 2
    let queuePriority = CLong(DISPATCH_QUEUE_PRIORITY_DEFAULT)
    let backgroundQueue = dispatch_get_global_queue(queuePriority,0)
    dispatch_async(backgroundQueue) {
      // 3
      let output = PHContentEditingOutput(contentEditingInput: self.input)
      // 4
      let archiveData = NSKeyedArchiver.archivedDataWithRootObject(self.currentFilterName!)
      let identifier = "com.raywenderlich.imgvue-imgvue-filters"
      let adjustmentData = PHAdjustmentData(formatIdentifier: identifier,
        formatVersion: "1.0", data: archiveData)
      output.adjustmentData = adjustmentData
      // 5
      if let path = self.input!.fullSizeImageURL.path {
        var image = UIImage(contentsOfFile: path)!
        image = self.filterService.applyFilter(self.currentFilterName!,
          toImage: image)
        // 6
        let jpegData = UIImageJPEGRepresentation(image, 1.0)
        var error: NSError?
        let saveSucceeded = jpegData.writeToURL(output.renderedContentURL,
          options: .DataWritingAtomic, error: &error)
        if saveSucceeded {
          // 7
          completionHandler(output)
        } else {
          // 8
          println("An error occurred during save: \(error)")
          completionHandler(nil)
        }
      } else {
        println("An error occurred loading the input image")
        completionHandler(nil)
      }
    }
  }
  
  var shouldShowCancelConfirmation: Bool {
    // Determines whether a confirmation to discard changes should be shown to the user on cancel.
    // (Typically, this should be "true" if there are any unsaved changes.)
    return false
  }
  
  func cancelContentEditing() {
    // Clean up temporary files, etc.
    // May be called after finishContentEditingWithCompletionHandler: while you prepare output.
  }
  
}
